clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -314,68 +369,92 @@ public void poll(long timeout) throws IOException {
         maybeCloseOldestConnection(endSelect);
     }
 
-    private void pollSelectionKeys(Iterable<SelectionKey> selectionKeys,
+    private void pollSelectionKeys(Set<SelectionKey> selectionKeys,
                                    boolean isImmediatelyConnected,
                                    long currentTimeNanos) {
-        Iterator<SelectionKey> iterator = selectionKeys.iterator();
-        while (iterator.hasNext()) {
-            SelectionKey key = iterator.next();
-            iterator.remove();
-            KafkaChannel channel = channel(key);
-
-            // register all per-connection metrics at once
-            sensors.maybeRegisterConnectionMetrics(channel.id());
-            if (idleExpiryManager != null)
-                idleExpiryManager.update(channel.id(), currentTimeNanos);
-
-            try {
-
-                /* complete any connections that have finished their handshake (either normally or immediately) */
-                if (isImmediatelyConnected || key.isConnectable()) {
-                    if (channel.finishConnect()) {
-                        this.connected.add(channel.id());
-                        this.sensors.connectionCreated.record();
-                        SocketChannel socketChannel = (SocketChannel) key.channel();
-                        log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
-                                socketChannel.socket().getReceiveBufferSize(),
-                                socketChannel.socket().getSendBufferSize(),
-                                socketChannel.socket().getSoTimeout(),
-                                channel.id());
-                    } else
-                        continue;
-                }
-
-                /* if channel is not ready finish prepare */
-                if (channel.isConnected() && !channel.ready())
-                    channel.prepare();
+        //it is possible that the iteration order over selectionKeys is the same every invocation.
+        //this may cause starvation of reads when memory is low. to address this we shuffle the keys if memory is low.
+        Collection<SelectionKey> inHandlingOrder;
+        List<SelectionKey> keysHandled = new ArrayList<>(selectionKeys.size());
+        if (!outOfMemory && memoryPool.availableMemory() < lowMemThreshold) {
+            List<SelectionKey> temp = new ArrayList<>(selectionKeys);
```

Comment:
Is there any concern about efficiency here? In particular, this allocates a new list and runs an additional linear time algorithm. A simpler alternative that has weaker randomization guarantees would be to select a random starting offset and use a couple of iterators to implement a sort of rotated view of the original list. Or perhaps overall it's not a concern since we have a linear cost to process all the keys anyway?